run/image-processing/e2e_test.py

```diff
@@ -0,0 +1,162 @@
+# Copyright 2020 Google, LLC.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This tests the Pub/Sub image processing sample
+
+import datetime
+import os
+import subprocess
+import time
+import uuid
+
+from google.cloud import storage
+from google.cloud.storage import Blob, notification
+
+import pytest
+
+
+SUFFIX = uuid.uuid4().hex[0:6]
+PROJECT = os.environ['GOOGLE_CLOUD_PROJECT']
+CLOUD_RUN_SERVICE = f"image-proc-{SUFFIX}"
+INPUT_BUCKET = f"image-proc-input-{SUFFIX}"
+OUTPUT_BUCKET = f"image-proc-output-{SUFFIX}"
+
+
+@pytest.fixture()
+def cloud_run_service():
+    # Build and Deploy Cloud Run Services
+    subprocess.run(
+        ["gcloud", "builds", "submit", "--project", PROJECT,
+         f"--substitutions=_SERVICE={CLOUD_RUN_SERVICE},_OUTPUT_BUCKET={OUTPUT_BUCKET}",
+         "--config=e2e_test_setup.yaml", "--quiet"],
+        check=True
+    )
+
+    yield
+
+    # Delete Cloud Run service and image container
+    subprocess.run(
+        ["gcloud", "run", "services", "delete", CLOUD_RUN_SERVICE,
+         "--project", PROJECT, "--platform", "managed", "--region",
+         "us-central1", "--quiet"],
+        check=True
+    )
+
+    subprocess.run(
+        ["gcloud", "container", "images", "delete",
+         f"gcr.io/{PROJECT}/{CLOUD_RUN_SERVICE}", "--quiet"],
+        check=True
+    )
+
+
+@pytest.fixture
+def service_url(cloud_run_service):
+    # Get the URL for the cloud run service
+    service_url = subprocess.run(
+        [
+            "gcloud",
+            "run",
+            "--project",
+            PROJECT,
+            "--platform=managed",
+            "--region=us-central1",
+            "services",
+            "describe",
+            CLOUD_RUN_SERVICE,
+            "--format=value(status.url)",
+        ],
+        stdout=subprocess.PIPE,
+        check=True
+    ).stdout.strip()
+
+    yield service_url.decode()
+
+
+@pytest.fixture()
+def pubsub_topic(service_url):
+    # Create pub/sub topic
+    topic = f"image_proc_{SUFFIX}"
+    subprocess.run(
```

Comment:
This is all really great, but in the future would it be better to this to be set up as a part of cloud build?

Reply:
Good question -- I didn't go down that road b/c the Cloud Build service agent didn't have pub/sub permissions in this project, but obviously we could update the permissions.